### ISSUE TITLE
Should Roslyn or SDK provide a mechanism to enable/disable some/all source generators?

### ISSUE THREAD

**Issue by ericstj** [dotnet/roslyn#55518]
User's might need to disable a source generator for some reason (like a performance or compatibility issue).  In general we should fix those and it shouldn't be a very common occurrence, however we may want to have a mechanism available as a mitigation.

In writing up the docs for https://github.com/dotnet/docs/pull/24896#discussion_r661102362 we came to the conclusion that allowing users to toggle all framework source generators/analyzers wasn't particularly useful, and might even be dangerous.

https://github.com/dotnet/sdk/issues/1212 means that analyzers/generators from packages can't easily be excluded either.

Should we have targets in the SDK/MSBuild that formalize a file-based exclusion mechanism?  Should roslyn support some exclusion mechanism itself (through compiler parameters, editorconfig, etc)?

cc @jaredpar @chsienki @terrajobst @KathleenDollard @jmarolf

**Comment by jcouv** [comment]
FYI @chsienki

**Comment by jmarolf** [comment]
I think the answer to this question is "Yes" just so that its easier for us to debug customers build issues.

**Comment by ericstj** [comment]
@maryamariyan and @sharwell were discussing this.  We'd like to see something land for 6.0 if possible.

**Comment by maryamariyan** [comment]
> @maryamariyan and @sharwell were discussing this. We'd like to see something land for 6.0 if possible.

We chatted about identifying a way to granularly disable source generators (VS2019 story for V1 source generator).

#### Key takeaways:
- Source generators can theoretically disable themselves in either of these two cases: (1) V2 Source generators or (2) V1 source generator which doesn't use a SyntaxReceiver.
- For V1 source generators with SyntaxReceiver, they have to be disabled in another way. Even if the generator is disabled the syntax receiver will run accounting for the slow part.
- For V2, we have other approaches to disabled (e.g. SuppressRazorSourceGenerator) but we may not need to consider disabling for V2 at all because there are no perf concerns there.

#### Suggested approach to disable for V1: msbuild logic
- For a given assembly that only contains the source generator, we make sure it recognizes a build property that when set, it removes it from source generators that can get used.

**Comment by maryamariyan** [comment]
@KathleenDollard @jaredpar @chsienki @terrajobst @jmarolf if you have thoughts against this proposed approach of using an msbuild property please let us know.

**Comment by jaredpar** [comment]
> We chatted about identifying a way to granularly disable source generators (VS2019 story for V1 source generator).

I can't tell if this proposal is for the generators owned by us (essentially first party generators) or if it's meant as a more general proposal for all generators. 

When you think about "all generators" there are other complications to consider. Mostly that customers can, and do, bundle multiple products into a single DLL. Namely a generator, analyzer and suppressor can all be in the same DLL. Hence the idea of a general mechanism being "remove the analyzer DLL from the build" is not going to work because removing a suppressor can introduce build breaks.

**Comment by jmarolf** [comment]
As we discussed in this comment thread https://github.com/dotnet/docs/pull/24896#discussion_r661102362 my position is we should have a `SkipGenerator` option that has an identical design to `SkipAnalyzers` from [here](https://github.com/dotnet/roslyn/blob/8c2dce8f7f30b33f18f8a58c3e4250f34f9ea9f3/src/Compilers/Core/MSBuildTask/Microsoft.Managed.Core.targets#L55). Generators are part of build correctness and I do not think any developer would have (or should be expected to have) the knowledge for how to disable them (much less selectively disable them).

No idea about the scheduling of this work but @jaredpar and @chsienki do you have any objections to this proposal?

**Comment by jaredpar** [comment]
My main objection to this is that it's possible this introduces build breaks. Consider that `/skipAnalyzers` used to skip both analyzers and suppressors. After massive amounts of feedback though we had to pull back and only skip analyzers. The reason being that many customers have warnings promoted to errors and were relying on suppressors to remove warnings. Hence by adding `/skipAnalyzers` new warnings popped up and that blocked the build.

The issue with `/skipGenerators` is that not all generators are optional like the JSON generator (allows for optimizations). Instead there are many, like the C# syntax generator, which are required for compilation to succeed. It's likely that many of them will end up falling into this bucket. Hence it seems this option is more likely than not going to cause compilation errors when invoked. 

Think there needs to be some sort of balance here for this to work. Essentially being able to differentiate between required generators and optional generators. Otherwise this flag won't be useful in practice

**Comment by jmarolf** [comment]
@jaredpar I agree that this is not as essential as /skipAnalyzers. essentially this is a diagnostic tool for us to use. I wish this was just an msbuild concept but there is not way to distinguish at an MSBuild level what a generator is from item metadata. 

> Think there needs to be some sort of balance here for this to work. Essentially being able to differentiate between required generators and optional generators. Otherwise this flag won't be useful in practice

This is a good insight. The main reason to implement `/skipAnalyzers` was to automatically elide analyzers in places where they were not needed for build performance. Perhaps we have some MSBuild convention where generators that are not required are added to a particular item group and not passed to the compiler if a property is present

**Comment by maryamariyan** [comment]
Seems like there is consensus on adding mechanism to disable all source generators, and to wait until there's further feedback for disabling individual generators.

**Comment by ericstj** [comment]
@jaredpar / @chsienki -- will this be part of the 6.0 / roslyn 4.0 release?  I think that's what the milestone 17.0 means but I wanted to confirm.  Thanks!

**Comment by jaredpar** [comment]
Yes VS 17.0 maps to .NET 6.0. The intent is for us to implement this during 17.0 Preview 5.

**Comment by Bosch-Eli-Black** [comment]
I'd like to see a way to mark generators as `required`;

This would allow us to have `/skipGenerators=optional`, `/skipGenerators=required`, and `/skipGenerators=all`. `/skipGenerators` could default to `/skipGenerators=optional`.

I think this would solve @jaredpar's concerns about builds being broken by disabling generators wholesale 🙂

**Comment by jaredpar** [comment]
> I'd like to see a way to mark generators as required.

My mental model of required is a generator that must run else the code will not compile. For example any generator that fills out a `partial` member. An optional generator is one that just improves the performance of existing scenarios. When not run the code has acceptable fallbacks that allow it to remain functional. 

The vast majority of generators fall into the required category. I can only think of 1-2 that actually fall into the optional category. As such I do not think we'd ever go down a path where we mark generators as required because that is the assumed default. Instead if we ever went down this path it would be to have generators be marked optional. 

> I think this would solve @jaredpar's concerns about builds being broken by disabling generators wholesale 🙂

I also want to note that it's been a year since this was filed and we have not found a concrete scenario for needing this flag. Yes we've had a few cases where generators have interferred with compilation: either via bad performance or crashing. In every case though we've been able to either correct the bug on our side or work with the generator author to correct it on their side.
